### PR TITLE
square-root implementation of continuous-time filtering

### PR DESCRIPTION
**Pull request by jzenn**
This adds a square-root implementation of continuous-time filtering to `LinearSDE` as described in Issue #258.

Some comments:

1. ~~I use `scipy.linalg.lu_factor` and `scipy.linalg.lu_solve` instead of two times `np.linalg.solve` or `np.linalg.inv` one time.~~
2. The covariance matrix of a random variable needs to be positive definite for the sqrt-algorithm to work. Currently there is no error-handling for a non-positive definite matrix.
3. There is no docstring for `forward_implementation` which I adapted from `DiscreteLinearGaussian`.
4. To avoid duplication I refactored `scipy.integrate.solve_ivp` into `_solve_mde_forward`.

### CODE REVIEW

**Comment by pnkraemer** [comment]
Hmmm, somehow this PR slipped past me, but it looks pretty ready! 

Let me know when you want a review please :)

**Comment by jzenn** [comment]
You can review the PR if you like :)

File: src/probnum/statespace/sde.py
Comment (pnkraemer) [inline src/probnum/statespace/sde.py:393]: Do you think it would be possible to document this a bit more extensively than linking the paper? It is paywall-protected, so not everyone can see the equation easily. One way would be to write down the actual equation (in a maths way) that is solved here. That would help me when reading the code :)

File: src/probnum/statespace/sde.py
Comment (pnkraemer) [inline src/probnum/statespace/sde.py:380]: isnt this sqrt a cholesky factor? if so, cho_solve would not only be faster but also more stable numerically

File: tests/test_statespace/test_sde.py
Comment (pnkraemer) [inline tests/test_statespace/test_sde.py:266]: Please correct me if I am wrong. The sqrt-forward implementation solves a different ODE than the non-square root implementation, both use an adaptive solver. So is it actually meaningful to check that the outputs are identical up to numerical precision? It is already tested whether they coincide with the LTI transition, right? 

I am not sure whether I am asking a question or proposing anything. But do you think that it would be a meaningful layer of testing to do the present test here on a non-lti sde?

File: src/probnum/statespace/sde.py
Comment (pnkraemer) [inline src/probnum/statespace/sde.py:251]: this block of code exists for both forward implementations, right? Do you think it is worth trying to factor this into a common implementation, or do you reckon it is wasted efforts?

**Review by pnkraemer — COMMENTED** [review]
Thanks a lot for addressing this! I left some comments in the code diff, nothing major. :)

File: src/probnum/statespace/sde.py
Comment (jzenn) [reply src/probnum/statespace/sde.py:251]: I thought about this but there would be a naming issue since once we get the square-root covariance and once the covariance (in 251). If that's ok for you I can change it.

File: src/probnum/statespace/sde.py
Comment (jzenn) [reply src/probnum/statespace/sde.py:393]: I can do that.

File: src/probnum/statespace/sde.py
Comment (pnkraemer) [reply src/probnum/statespace/sde.py:251]: Hmmm I dont know. I am a bit worried about all the flattening and slicing that is happening here, so in case some error comes up, it would be good to only have to fix it in one place. I see the naming though, and perhaps it is actually more readable if it is *not* factored out. As long as it only exists, like, twice, I am fine with a bit of duplication. 

Long story short, I dont know -- your call!

Comment (jzenn) [reply src/probnum/statespace/sde.py:251]: Would something like `new_cov_or_cov_sqrt` resolve the naming issue?

Comment (pnkraemer) [reply src/probnum/statespace/sde.py:251]: I think it would help, especially with an accompanying comment that explains why the `or`. Not sure whether it entirely resolves the issue though...

**Comment by pnkraemer** [comment]
Can you also add the `forward_implementation` argument to `ContinuousEKFComponent` please? By implementing forward and backward transitions for `LinearSDE` objects, they are immediately available for continuous-time extended Kalman filtering (ContinuousUKF does not work quite the same with the current code, unfortunately). 

A guide is the respective choice of implementation in `DiscreteEKFComponent`. 

Thanks! :)

File: src/probnum/statespace/sde.py
Comment (jzenn) [reply src/probnum/statespace/sde.py:380]: I initially thought that we cannot rely on `cov_sqrt` being positive definite but you're right it is initialised using `initrv.cov_cholesky`.

File: tests/test_statespace/test_sde.py
Comment (jzenn) [reply tests/test_statespace/test_sde.py:266]: This should test the forward values being the same for both implementations. I thought for the simple LTI case they should coincide. Nevertheless, I think testing on a non-LTI SDE might be a good idea.

**Comment by jzenn** [comment]
> Can you also add the `forward_implementation` argument to `ContinuousEKFComponent` please? By implementing forward and backward transitions for `LinearSDE` objects, they are immediately available for continuous-time extended Kalman filtering (ContinuousUKF does not work quite the same with the current code, unfortunately).

I haven't thought about this. I will add `forward_implementation`.

File: tests/test_statespace/test_sde.py
Comment (jzenn) [reply tests/test_statespace/test_sde.py:266]: > It is already tested whether they coincide with the LTI transition, right?
If I understand you correctly, it is already tested whether the classic implementation coincides with the LTI transition but not whether the square-root implementation coincides with the LTI transition.

Comment (pnkraemer) [reply tests/test_statespace/test_sde.py:266]: Ah I see. In this case nevermind my comment :) Thanks!

File: src/probnum/statespace/sde.py
Comment (jzenn) [reply src/probnum/statespace/sde.py:380]: I just looked into this again realizing that I did not use a Cholesky decomposition since `cov_sqrt` is not symmetric. Sorry for the confusion!

**Comment by jzenn** [comment]
I just realized that `driftmatfun` is called `G(t)` in the docstring while the `_setup_*` methods call it `F(t)`. I will change that to be consistent.

File: src/probnum/statespace/sde.py
Comment (pnkraemer) [inline src/probnum/statespace/sde.py:393]: wow, thanks a lot for this! Do you think it would help to put these maths environments into a `.. math:: \dot S(...` environment? This would make sphinx render this docstring well (which currently is not needed, because this function is not visible). I dont know whether it would be worth the effort. What do you reckon?

File: src/probnum/statespace/sde.py
Comment (pnkraemer) [inline src/probnum/statespace/sde.py:415]: the maths above would suggest to me that cov_sqrt ("S") is triangular (because of this choice of $M$, or am I misunderstanding?

File: src/probnum/statespace/sde.py
Comment (pnkraemer) [inline src/probnum/statespace/sde.py:255]: I think one can also initialise the `cov_cholesky` field of `Normal` with `new_cov_sqrt` (after all, that was the point here, right?). Or am I misunderstanding?

**Review by pnkraemer — COMMENTED** [review]
Thanks a lot for your efforts johannes! I had some questions regarding the cov_sqrt (and whether it is a Cholesky factor, because if it is, we may want to use this at some places). I left those in the code diff :)

File: src/probnum/statespace/sde.py
Comment (jzenn) [reply src/probnum/statespace/sde.py]: @pnkraemer here

File: src/probnum/statespace/sde.py
Comment (jzenn) [reply src/probnum/statespace/sde.py:415]: Yes you're right. I left a comment above in the last review explaining this.

File: src/probnum/statespace/sde.py
Comment (jzenn) [reply src/probnum/statespace/sde.py:393]: I can do that

File: src/probnum/statespace/sde.py
Comment (jzenn) [reply src/probnum/statespace/sde.py:255]: Yes, I totally missed this.

File: src/probnum/statespace/sde.py
Comment (jzenn) [reply src/probnum/statespace/sde.py:415]: I will change that to `scipy.linalg.solve_triangular`.

File: src/probnum/filtsmooth/gaussfiltsmooth/extendedkalman.py
Comment (pnkraemer) [inline src/probnum/filtsmooth/gaussfiltsmooth/extendedkalman.py:139]: thanks for doing this! Is this argument in the docstring?

File: src/probnum/statespace/sde.py
Comment (pnkraemer) [inline src/probnum/statespace/sde.py:357]: riccati, not ricatti (right?)

File: src/probnum/statespace/sde.py
Comment (pnkraemer) [inline src/probnum/statespace/sde.py:249]: would be great (if this is a cholesky factor) to call this `sol_cov_cholesky` (in line with how we refer to this variable in e.g. `Normal`).

Comment (pnkraemer) [reply src/probnum/statespace/sde.py:249]: this applies to all the code in this module :)

File: src/probnum/statespace/sde.py
Comment (pnkraemer) [inline src/probnum/statespace/sde.py:277]: thanks for these explanations :)

File: src/probnum/statespace/sde.py
Comment (pnkraemer) [inline src/probnum/statespace/sde.py:418]: `cov_cholesky` would be better to have here, because it conveys more information than `cov_sqrt`

**Review by pnkraemer — COMMENTED** [review]
Thank you so much! I am sorry that there have been so many iterations with this PR, and that it has been pending for a while now. Once my final comments (all are super minor) are resolved, let's merge it!

File: src/probnum/statespace/sde.py
Comment (jzenn) [reply src/probnum/statespace/sde.py:357]: Yes, you're right :)

File: src/probnum/filtsmooth/gaussfiltsmooth/extendedkalman.py
Comment (jzenn) [reply src/probnum/filtsmooth/gaussfiltsmooth/extendedkalman.py:139]: No, it also is not for `LinearSDE` and `DiscreteLinearGaussian` (see also 3. in https://github.com/probabilistic-numerics/probnum/pull/374#issue-610816053).

File: src/probnum/statespace/sde.py
Comment (jzenn) [reply src/probnum/statespace/sde.py:249]: >if this is a cholesky factor

The paper does not explicitly state anything regarding `cov_sqrt` being a Cholesky factor. I think `cov_sqrt` should be a Cholesky factor but I am not certain.

File: src/probnum/filtsmooth/gaussfiltsmooth/extendedkalman.py
Comment (pnkraemer) [reply src/probnum/filtsmooth/gaussfiltsmooth/extendedkalman.py:139]: Hm okay, I can live with that :)

File: src/probnum/statespace/sde.py
Comment (pnkraemer) [reply src/probnum/statespace/sde.py:249]: I think it should (because of the way it is truncated after each step), but you know more about this than I do :) One simple way of finding out would be to run the tests with a simple `assert np.testing.assert_allclose(cov_sqrt, np.tril(cov_sqrt))` and if it passes, it should be a cholesky factor. If so, then renaming these variables would be wonderful, and if not, leave it as is (I am truly sorry to be so petty here with respect to the naming, but this information whether it is lower triangular or not is fairly crucial IMO -- after this we can merge, I promise!! :) )

Comment (jzenn) [reply src/probnum/statespace/sde.py:249]: >One simple way of finding out would be to run the tests with a simple assert np.testing.assert_allclose(cov_sqrt, np.tril(cov_sqrt)) and if it passes, it should be a cholesky factor

Yes, this is true by definition. However, I don't see that `cov_sqrt_{ii} > 0` which does need to be true for a Cholesky factor, right?

File: src/probnum/filtsmooth/gaussfiltsmooth/extendedkalman.py
Comment (jzenn) [reply src/probnum/filtsmooth/gaussfiltsmooth/extendedkalman.py:139]: I can introduce the argument in the docstring if you'd feel more comfortable with that :)

File: src/probnum/statespace/sde.py
Comment (pnkraemer) [reply src/probnum/statespace/sde.py:249]: Ah I see. You can transform the factor into a valid cholesky factor though: `S = L1 L1t = L2 Q Qt L2t` and if `Qt` is the (orthogonal) matrix consisting of 1 and -1 on the diagonals, then (in this context here) we can rather loosely switch back between L1 and L2 in order to transform the lower-triangular matrix-sqrt into a Cholesky factor (its diagonals need to be != 0, but that should be no problem here). This is what is done in `probnum.utils.linalg.cholesky_update`, and might be applicable here. What do you think?

Comment (jzenn) [reply src/probnum/statespace/sde.py:249]: Oh I see. I haven't thought on that end. My tests suggested that the diagonal entries are indeed positive but I did not see a way to proof this following the algorithm of the paper. I will look into `probnum.utils.linalg.cholesky_update` and report back if that is ok.

**Comment by jzenn** [comment]
One more question regarding `_cholesky` and `_sqrt`: Would you recommend renaming the methods as well (and the class parameter)?

**Comment by pnkraemer** [comment]
no need, the `_sqrt`, `_classic` etc. endings are used everywhere around here, and are arguably easier to parse (the algorithm is called square-root kalman filter) :)

File: src/probnum/statespace/sde.py
Comment (pnkraemer) [reply src/probnum/statespace/sde.py:250]: I see that you re-requested my review -- may I ask what the result here was? :)

Comment (jzenn) [reply src/probnum/statespace/sde.py:250]: I am using `probnum.utils.linalg.cholesky_update` to ensure that `cov_cholesky` indeed is a Cholesky factor as you suggested. However, I did not come up with an example where the algorithm produces a triangular matrix with non-positive diagonal entries.

**Comment by jzenn** [comment]
Is there anything else I can do here @pnkraemer?

File: src/probnum/statespace/sde.py
Comment (pnkraemer) [inline src/probnum/statespace/sde.py:437]: I think you meant `utils.linalg.tril_to_positive_tril` here?!

**Review by pnkraemer — COMMENTED** [review]
Hey, sorry for the delay again. I think the call to Cholesky update is mixed up (it should rather be `tril_to_positive_tril`, and the branch is not up to date anymore. I promise to approve when these two things are resolved!

File: src/probnum/statespace/sde.py
Comment (jzenn) [reply src/probnum/statespace/sde.py:437]: I think this should result in the same matrices, right? But I see that using `utils.linalg.tril_to_positive_tril` might be more descriptive of this use case (and we'd save the call to `np.linalg.qr`).

Comment (pnkraemer) [reply src/probnum/statespace/sde.py]: yes, the call to qr is the things that should be avoided here. QR decompositions are expensive :)

**Review by pnkraemer — APPROVED** [review]
Thanks a lot for your patience johannes!

**Pull request merged**
